pkg/koordlet/metricsadvisor/performance_collector_linux.go

```diff
@@ -82,26 +83,35 @@ func (c *performanceCollector) collectContainerCPI() {
 
 	time.Sleep(time.Duration(c.collectTimeWindow) * time.Second)
 
+	// collect cpi, psi for each container at the same time
 	var wg1 sync.WaitGroup
 	wg1.Add(len(containerStatusesMap))
 	for containerStatus, podMeta := range containerStatusesMap {
 		podUid := podMeta.Pod.UID
-		go func(status *corev1.ContainerStatus, podUid string) {
+		cgroupDir := podMeta.CgroupDir
+		go func(parentDir string, status *corev1.ContainerStatus, podUid string) {
 			defer wg1.Done()
+			// collect container cpi
 			oneCollector, ok := collectors.Load(status)
 			if !ok {
 				return
 			}
-			c.profilePerfOnSingleContainer(status, oneCollector.(*perf.PerfCollector), podUid)
-
-			err1 := unix.Close(oneCollector.(*perf.PerfCollector).CgroupFd)
+			perfCollector, ok := oneCollector.(*perf.PerfCollector)
+			if !ok {
+				klog.Errorf("PerfCollector type convert failed")
+				return
+			}
+			c.profilePerfOnSingleContainer(status, perfCollector, podUid)
+			err1 := perfCollector.CleanUp()
 			if err1 != nil {
-				klog.Errorf("close CgroupFd %v, err : %v", oneCollector.(*perf.PerfCollector).CgroupFd, err1)
+				klog.Errorf("PerfCollector cleanup err : %v", err1)
 			}
-		}(containerStatus, string(podUid))
+			// collect container psi
+			c.collectSingleContainerPSI(parentDir, status, podUid)
```

Comment:
psi collector should be an independent procedure, should be moved out of cpi collection.
the interval can be short since the overhead of collection is low